designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,531 @@
+    yield* this._cliEngine.executeOnFiles(patterns).results
+  }
+}
+```
+
+But once [RFC42] is implemented, the returned object will be an instance of [`LintResultGenerator`](https://github.com/eslint/eslint/blob/836c0e48704d70bc1a5cbdbf0211368b0ada942d/lib/eslint/lint-result-generator.js#L136) class.
+
+</details>
+
+If you want to use the returned object with `await` expression, you can use a small utility to convert an async iterable object to an array.
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+// Convert the results to an array.
+const results = await ESLint.collectResults(eslint.executeOnFiles(patterns))
+
+// Optionally you can sort the results.
+results.sort(ESLint.compareResultsByFilePath)
+
+print(results)
+```
+
+Once we got this change, we can realize the following things:
+
+- [RFC42] We can implement linting in parallel by worker threads. It will reduce spending time of linting much.
+- [RFC45] We can implement to print the results in streaming or to print progress state, without more breaking changes. Because ESLint may spend time to lint files (for example, ESLint needs about 20 seconds to lint [our codebase](https://github.com/eslint/eslint)), to print progress state will be useful.
+- (no RFC yet) We can support ES modules for shareable configs, plugins, and custom parsers.
+
+##### Iterate only one time
+
+We can iterate the returned object of this method only one time similar to generators.
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+const resultGenerator = eslint.executeOnFiles(patterns)
+for await (const result of resultGenerator) {
+  print(result)
+}
+// ↓ Throw "This generator has been consumed already"
+for await (const result of resultGenerator) {
+  print(result)
+}
+```
+
+##### Move the `usedDeprecatedRules` property
+
+The returned object of `CLIEngine#executeOnFiles()` has the `usedDeprecatedRules` property that includes the deprecated rule IDs which the linting used.
+
+But the location doesn't fit asynchronous because the used deprecated rule list is not determined until the iteration finished. Therefore, this RFC moves the `usedDeprecatedRules` property to each lint result.
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+
+for await (const result of eslint.executeOnFiles(patterns)) {
+  console.log(result.usedDeprecatedRules)
+}
+```
+
+As a side-effect, formatters gets the capability to print the used deprecated rules. Previously, ESLint has not passed the returned object to formatters, so the formatters could not print used deprecated rules. After this RFC, each lint result has the `usedDeprecatedRules` property and the formatters receive those.
+
+##### Disallow execution in parallel
+
+Because this method updates the cache file, it will break the cache file if called multiple times in parallel. To prevent that, every call of `executeOnFiles()` must wait for the previous call finishes.
```

Comment:
This seems like a pretty big constraint to place on API consumers. It seems like each run should be able to:

1. Read data in from the cache file, store the mtime or hash
2. Do its linting
3. Write data back to the cache file if the mtime or hash is the same

Of course, this would only be if the run makes use of a cache file, but I don't think throwing an error when a second run starts while a first run is in process makes sense. We don't know how many instances of `ESLint` (or whatever the name ends up being) an application will create.

Reply:
Oh, I'm sorry for unclear. This is not a constraint for API customers. This is a constraint for the implementation of this method. If a user called this method twice in parallel, the latter waits for the former internally.

> We don't know how many instances of ESLint (or whatever the name ends up being) an application will create.

True. So should we up this exclusive control to users completely?

Reply:
> 1. Read data in from the cache file, store the mtime or hash
> 1. Do its linting
> 1. Write data back to the cache file if the mtime or hash is the same

Sounds reasonable. I will update this section.